d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java

```diff
@@ -302,19 +276,138 @@ private ClusterProperties toClusterProperties(Struct clusterProperties, long ver
+      parsedMap.put(entry.getKey(), uriProperties);
+    }
+
+    return parsedMap;
+  }
+
+  private class UriPropertiesResourceWatcher implements XdsClient.D2NodeMapResourceWatcher
+  {
+    final String _clusterName;
+    final AtomicBoolean _init;
+    final long _initFetchStart;
+
+    Map<String, UriProperties> _currentData = new HashMap<>();
+
+    public UriPropertiesResourceWatcher(String clusterName)
+    {
+      _clusterName = clusterName;
+      _init = new AtomicBoolean();
+      _initFetchStart = System.nanoTime();
+    }
+
+    @Override
+    public void onChanged(XdsClient.D2NodeMapUpdate update)
+    {
+      if (_init.compareAndSet(false, true))
+      {
+        emitSDStatusInitialRequestEvent(_clusterName);
+      }
+
+      if (_uriEventBus != null)
+      {
+        try
+        {
+          Map<String, UriProperties> updates = toUriProperties(update.getNodeDataMap());
+          emitSDStatusUpdateReceiptEvents(updates);
+          _currentData = updates;
+          UriProperties mergedUriProperties = _uriPropertiesMerger.merge(_clusterName, _currentData.values());
+          _uriEventBus.publishInitialize(_clusterName, mergedUriProperties);
+          if (_dualReadStateManager != null)
+          {
+            _dualReadStateManager.reportData(_clusterName, mergedUriProperties, true);
+          }
+        }
+        catch (InvalidProtocolBufferException | PropertySerializationException e)
+        {
+          _log.error("Failed to parse D2 uri properties from xDS update. Cluster name: " + _clusterName, e);
+        }
+      }
+    }
+
+    @Override
+    public void onError(Status error)
+    {
+      notifyAvailabilityChanges(false);
+    }
+
+    @Override
+    public void onReconnect()
+    {
+      notifyAvailabilityChanges(true);
+    }
+
+    private void emitSDStatusInitialRequestEvent(String cluster)
+    {
+      if (_eventEmitter == null)
+      {
+        _log.info("Service discovery event emitter in ZookeeperEphemeralStore is null. Skipping emitting events.");
+        return;
+      }
+
+      // measure request duration and convert to milli-seconds
+      long initialFetchDurationMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - _initFetchStart);
+      if (initialFetchDurationMillis < 0)
+      {
+        _log.warn("Failed to log ServiceDiscoveryStatusInitialRequest event, initialFetchStartAt time is greater than current time.");
+        return;
+      }
+      // emit service discovery status initial request event for success
+      _eventEmitter.emitSDStatusInitialRequestEvent(cluster, true, initialFetchDurationMillis, true);
+
     }
 
-    return _uriPropertiesMerger.merge(clusterName, allUriProperties);
+    private void emitSDStatusUpdateReceiptEvents(Map<String, UriProperties> updates)
+    {
+      if (_eventEmitter == null)
+      {
+        _log.info("Service discovery event emitter in ZookeeperEphemeralStore is null. Skipping emitting events.");
```

Comment:
ZookeeperEphemeralStore should be XdsToD2PropertiesAdaptor, same applies to the other check for initial request.